spark3/src/main/java/org/apache/iceberg/spark/MigrateAction.java

```diff
@@ -0,0 +1,272 @@
+ *   import has succeeded.
+ *
+ *   - Replacing a table in the Session Catalog with an Iceberg Table with the same name
+ *   This pathway will create a temporary table with a different name. This replacement table will
+ *   be committed upon a successful import. Then the original session catalog entry will be dropped
+ *   and the new replacement table renamed to take its place.
+ */
+public class MigrateAction implements Action<Long> {
+  private static final Logger LOG = LoggerFactory.getLogger(ExpireSnapshotsAction.class);
+  private static final Set<String> ALLOWED_SOURCES = ImmutableSet.of("parquet", "avro", "orc", "hive");
+  private static final String ICEBERG_METADATA_FOLDER = "metadata";
+  private static final String REPLACEMENT_NAME = "_REPLACEMENT_";
+
+
+  private final SparkSession spark;
+
+  // Source Fields
+  private final CatalogTable sourceTable;
+  private final String sourceTableLocation;
+  private final Identifier sourceTableName;
+  private final PartitionSpec sourcePartitionSpec;
+
+  // Destination Fields
+  private final Boolean sessionCatalogReplacement;
+  private final SparkSessionCatalog destCatalog;
+  private final Identifier destTableName;
+
+  // Optional Parameters for destination
+  private String destDataLocation;
+  private String destMetadataLocation;
+  private Map<String, String> additionalProperties = Maps.newHashMap();
+
+  /**
+   * Creates an Iceberg Location at a given location instead of using the location
+   * provided by the source table. New metadata and data files will be added to this
+   * new location and further operations will not effect the source table.
+   *
+   * Use this if you would like to experiment with Iceberg without changing
+   * your original files.
+   * @param newLocation the base directory for the new Iceberg Table
+   * @return this for chaining
+   */
+  public MigrateAction withNewTableLocation(String newLocation) {
+    this.destDataLocation = newLocation;
+    this.destMetadataLocation = newLocation + "/" + ICEBERG_METADATA_FOLDER;
+    return this;
+  }
+
+  /**
+   * Adds additional properties to the newly created Iceberg Table. Any properties with
+   * the same key name will be overwritten.
+   * @param properties a map of properties to be included
+   * @return this for chaining
+   */
+  public MigrateAction withAdditionalProperties(Map<String, String> properties) {
+    this.additionalProperties.putAll(properties);
+    return this;
+  }
+
+  /**
+   * Adds an additional property to the newly created Iceberg Table. Any properties
+   * with the same key name will be overwritten.
+   * @param key the key of the property to add
+   * @param value the value of the property to add
+   */
+  public MigrateAction withAdditionalProperty(String key, String value) {
+    this.additionalProperties.put(key, value);
+    return this;
+  }
+
+  public MigrateAction(SparkSession spark, Identifier destTableName, Identifier sourceTableName) {
```

Comment:
We normally put constructors before public methods. Minor, but this was harder to find than needed.